File: src-tauri/src/commands/collections.rs
// http_client/src-tauri/src/commands/collections.rs
use tauri::State;

use crate::commands::dto::{
    CollectionContentsDto, CollectionDto, ExampleDto, SaveExampleInput, SavedRequestDto,
    SendRequestInput,
};
use crate::commands::error::ApiError;
use crate::domain::models::{HttpRequest, KeyValue};
use crate::AppState;

/// Every command here is an adapter: parse, delegate, map. SQLite calls are
/// blocking, so they run on a blocking task like the HTTP ones do.
macro_rules! blocking {
    ($state:expr, $service:ident, $body:expr) => {{
        let $service = $state.collections.clone();
        tauri::async_runtime::spawn_blocking(move || $body)
            .await
            .map_err(|_| ApiError::internal("storage task failed to complete"))?
    }};
}

#[tauri::command]
pub async fn list_collections(state: State<'_, AppState>) -> Result<Vec<CollectionDto>, ApiError> {
    let collections = blocking!(state, service, service.list())?;
    Ok(collections.into_iter().map(CollectionDto::from).collect())
}

#[tauri::command]
pub async fn collection_contents(
    state: State<'_, AppState>,
    collection_id: String,
) -> Result<CollectionContentsDto, ApiError> {
    let contents = blocking!(state, service, service.contents(&collection_id))?;
    Ok(contents.into())
}

#[tauri::command]
pub async fn create_collection(
    state: State<'_, AppState>,
    name: String,
) -> Result<CollectionDto, ApiError> {
    let collection = blocking!(state, service, service.create_collection(&name))?;
    Ok(collection.into())
}

#[tauri::command]
pub async fn rename_collection(
    state: State<'_, AppState>,
    id: String,
    name: String,
) -> Result<(), ApiError> {
    blocking!(state, service, service.rename_collection(&id, &name))?;
    Ok(())
}

#[tauri::command]
pub async fn delete_collection(state: State<'_, AppState>, id: String) -> Result<(), ApiError> {
    blocking!(state, service, service.delete_collection(&id))?;
    Ok(())
}

#[tauri::command]
pub async fn create_folder(
    state: State<'_, AppState>,
    collection_id: String,
    parent_folder_id: Option<String>,
    name: String,
) -> Result<crate::commands::dto::FolderDto, ApiError> {
    let folder = blocking!(
        state,
        service,
        service.create_folder(&collection_id, parent_folder_id.as_deref(), &name)
    )?;
    Ok(folder.into())
}

#[tauri::command]
pub async fn rename_folder(
    state: State<'_, AppState>,
    id: String,
    name: String,
) -> Result<(), ApiError> {
    blocking!(state, service, service.rename_folder(&id, &name))?;
    Ok(())
}

#[tauri::command]
pub async fn delete_folder(state: State<'_, AppState>, id: String) -> Result<(), ApiError> {
    blocking!(state, service, service.delete_folder(&id))?;
    Ok(())
}

#[tauri::command]
pub async fn save_request(
    state: State<'_, AppState>,
    id: Option<String>,
    collection_id: String,
    folder_id: Option<String>,
    name: String,
    request: SendRequestInput,
) -> Result<SavedRequestDto, ApiError> {
    let domain_request = HttpRequest::try_from(request)?;
    let saved = blocking!(
        state,
        service,
        service.save_request(
            id,
            &collection_id,
            folder_id.as_deref(),
            &name,
            domain_request
        )
    )?;
    Ok(saved.into())
}

#[tauri::command]
pub async fn load_request(
    state: State<'_, AppState>,
    id: String,
) -> Result<SavedRequestDto, ApiError> {
    let saved = blocking!(state, service, service.load_request(&id))?;
    Ok(saved.into())
}

#[tauri::command]
pub async fn rename_request(
    state: State<'_, AppState>,
    id: String,
    name: String,
) -> Result<(), ApiError> {
    blocking!(state, service, service.rename_request(&id, &name))?;
    Ok(())
}

#[tauri::command]
pub async fn move_request(
    state: State<'_, AppState>,
    id: String,
    folder_id: Option<String>,
) -> Result<(), ApiError> {
    blocking!(
        state,
        service,
        service.move_request(&id, folder_id.as_deref())
    )?;
    Ok(())
}

#[tauri::command]
pub async fn delete_request(state: State<'_, AppState>, id: String) -> Result<(), ApiError> {
    blocking!(state, service, service.delete_request(&id))?;
    Ok(())
}

#[tauri::command]
pub async fn save_example(
    state: State<'_, AppState>,
    example: SaveExampleInput,
) -> Result<ExampleDto, ApiError> {
    let request = HttpRequest::try_from(example.request)?;
    let headers: Vec<KeyValue> = example
        .response_headers
        .into_iter()
        .map(KeyValue::from)
        .collect();
    let saved = blocking!(
        state,
        service,
        service.save_example(
            &example.request_id,
            &example.name,
            request,
            example.status,
            headers,
            example.response_body
        )
    )?;
    Ok(saved.into())
}

#[tauri::command]
pub async fn load_example(state: State<'_, AppState>, id: String) -> Result<ExampleDto, ApiError> {
    let example = blocking!(state, service, service.load_example(&id))?;
    Ok(example.into())
}

#[tauri::command]
pub async fn rename_example(
    state: State<'_, AppState>,
    id: String,
    name: String,
) -> Result<(), ApiError> {
    blocking!(state, service, service.rename_example(&id, &name))?;
    Ok(())
}

#[tauri::command]
pub async fn delete_example(state: State<'_, AppState>, id: String) -> Result<(), ApiError> {
    blocking!(state, service, service.delete_example(&id))?;
    Ok(())
}

File: src/services/collections.ts
// http_client/src/services/collections.ts
//
// One function per command in src-tauri/src/commands/collections.rs. The
// boundary crossing itself lives in services/invoke.ts, which is the only
// place invoke() appears for these (CLAUDE.md section 11, rule 3) — see
// src/types/collections.ts for the DTOs and the note in types/http.ts about
// ApiError's notFound/storage kinds, which these commands are the ones that
// actually produce.
import { call } from "@/services/invoke";
import type {
  Collection,
  CollectionContents,
  Example,
  Folder,
  SavedRequest,
  SavedWebSocket,
} from "@/types/collections";
import type { ApiError, KeyValue, SendRequestInput } from "@/types/http";
import type { Result } from "@/types/result";
import type { WebSocketRequest, WsDraft } from "@/types/websocket";

export function listCollections(): Promise<Result<Collection[], ApiError>> {
  return call<Collection[]>("list_collections");
}

export function collectionContents(
  collectionId: string,
): Promise<Result<CollectionContents, ApiError>> {
  return call<CollectionContents>("collection_contents", { collectionId });
}

export function createCollection(name: string): Promise<Result<Collection, ApiError>> {
  return call<Collection>("create_collection", { name });
}

export function renameCollection(id: string, name: string): Promise<Result<void, ApiError>> {
  return call<void>("rename_collection", { id, name });
}

export function deleteCollection(id: string): Promise<Result<void, ApiError>> {
  return call<void>("delete_collection", { id });
}

export function createFolder(
  collectionId: string,
  parentFolderId: string | null,
  name: string,
): Promise<Result<Folder, ApiError>> {
  return call<Folder>("create_folder", { collectionId, parentFolderId, name });
}

export function renameFolder(id: string, name: string): Promise<Result<void, ApiError>> {
  return call<void>("rename_folder", { id, name });
}

export function deleteFolder(id: string): Promise<Result<void, ApiError>> {
  return call<void>("delete_folder", { id });
}

export interface SaveRequestArgs {
  /** null saves a new request; an existing id overwrites it. */
  id: string | null;
  collectionId: string;
  folderId: string | null;
  name: string;
  request: SendRequestInput;
}

export function saveRequest(args: SaveRequestArgs): Promise<Result<SavedRequest, ApiError>> {
  return call<SavedRequest>("save_request", {
    id: args.id,
    collectionId: args.collectionId,
    folderId: args.folderId,
    name: args.name,
    request: args.request,
  });
}

export function loadRequest(id: string): Promise<Result<SavedRequest, ApiError>> {
  return call<SavedRequest>("load_request", { id });
}

export function renameRequest(id: string, name: string): Promise<Result<void, ApiError>> {
  return call<void>("rename_request", { id, name });
}

export function moveRequest(id: string, folderId: string | null): Promise<Result<void, ApiError>> {
  return call<void>("move_request", { id, folderId });
}

export function deleteRequest(id: string): Promise<Result<void, ApiError>> {
  return call<void>("delete_request", { id });
}

export interface SaveWebSocketArgs {
  /** null saves a new WebSocket request; an existing id overwrites it. */
  id: string | null;
  collectionId: string;
  folderId: string | null;
  name: string;
  request: WebSocketRequest;
  draft: WsDraft;
}

/** Rename, move, delete and docs for a WebSocket use the request functions
 * above: the commands act on a row by id, whatever its kind. */
export function saveWebSocket(args: SaveWebSocketArgs): Promise<Result<SavedWebSocket, ApiError>> {
  return call<SavedWebSocket>("save_web_socket", { input: args });
}

export function loadWebSocket(id: string): Promise<Result<SavedWebSocket, ApiError>> {
  return call<SavedWebSocket>("load_web_socket", { id });
}

export interface SaveExampleArgs {
  requestId: string;
  name: string;
  /** The request as sent, already resolved. */
  request: SendRequestInput;
  status: number;
  responseHeaders: KeyValue[];
  responseBody: string;
}

export function saveExample(args: SaveExampleArgs): Promise<Result<Example, ApiError>> {
  return call<Example>("save_example", { example: args });
}

export function loadExample(id: string): Promise<Result<Example, ApiError>> {
  return call<Example>("load_example", { id });
}

export function renameExample(id: string, name: string): Promise<Result<void, ApiError>> {
  return call<void>("rename_example", { id, name });
}

export function deleteExample(id: string): Promise<Result<void, ApiError>> {
  return call<void>("delete_example", { id });
}
